### PR TITLE
feat: CLI event hook flags

### PR DESCRIPTION
**Pull request by Harris-Miller**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [ ] will soon
- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

#### Overview

When in watch mode, it is not uncommon to want to kick off other tasks at the start or end of a bundle. The built-in CLI already has callbacks set up for the 5 events. This MR sets up config and cli flags to be able to set up custom hooks to run on those events in addition to the current operations.

This MR is incomplete on purpose. Before writing tests and updating documentation, I wanted to see if this feature is something the core team would like. See below for list of TODOs

#### Motivation

I work I use `yalc` heavily when developing locally. We use `rollup` as well. Right now are running a separate `rollup -c --watch` and `chokidar dist -c 'yalc push'` to marry incremental builds from `rollup` and moving the results of that build to other libs via `yalc`.

The `chokidar` watch is not necessary though because `rollup` has event hooks for `'START'`, `'END'`, etc. Except the built-in CLI currently only uses those hooks to display build information. It would be great to run `yalc push`, or any terminal command for that matter, at any of these hook points via the built-in CLI

#### Current Implementation

Using the systems already in place, I added the following CLI flag:
```
--watch.onError
--watch.onStart
--watch.onStartBundle
--watch.onEndBundle
--watch.onEnd
```

Each value is passed to `child_process.execSync`. To be run. `stdio` is set to `inherit` unless `--silent` is set, then it is `ignore`. I choose to use `execSync`, but I'm not opposed to using `exec` or `spawn` either, whatever the rollup team would like, just let me know what would be preferred here.

I specifically choose for these to be CLI options only since the intention is to add terminal commands to the built-in CLI. These would never apply to `WatchOptions` passed in API, eg `const watcher = rollup.watch(watchOptions);`, as users writing scripts utilizing the API would implement their own `watch.on('event')` to do what they needed

#### TODO

As mentioned before, I wanted to gauge if the rollup team would be open to accepting this MR before I completed it

- [x] Tests
- [x] Documentation

#### Final Thoughts

Thank you for the consideration of this MR. IMHO this feature is the main thing missing from the CLI. The level of effort it too to add this feature was minimal, yet it adds a lot of power to users.

### CODE REVIEW

**Review by lukastaegert — COMMENTED** [review]
Actually I think this is an interesting idea with minimal impact. If you add tests and documentation, I would merge this. `execSync` sounds fine to me as I think this would help the expectation that the command can be executed at exactly the point in time where the event happens, and we also do not expect unrelated tasks to share the event loop.

**Comment by Harris-Miller** [comment]
[lukastaegert](https://github.com/lukastaegert) Cool cool. I'm going to try and carve out some time next weekend or the weekend after to follow up on this MR and add tests and documentation. I'll ping you when that work is done for you to re-review. Thank you!

File: test/cli/samples/watch/watch-event-hooks/_config.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+		'rollup -cw --watch --watch.onStart "echo start" --watch.onBundleStart "echo bundleStart" --watch.onBundleEnd "echo bundleEnd" --watch.onEnd "echo onEnd" --watch.onError "echo onError"',
+	stderr(stderr) {
+		assertIncludes(
+			stderr,
+			`watch.onStart $ echo start
+start
+bundles main.js → _actual...
+watch.onBundleStart $ echo bundleStart
+bundleStart
+created _actual in 16ms
```
Comment (lukastaegert) [inline test/cli/samples/watch/watch-event-hooks/_config.js:15]: A CLI test must not depend on things like runtime. The usual solution is to use `assertIncludes` for substrings like `created _actual` etc.

File: test/cli/samples/watch/watch-event-hooks/_config.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const { assertIncludes } = require('../../../../utils.js');
+
+module.exports = {
+	description: 'onStart event hoot shell command executes correctly',
+	command:
+		'rollup -cw --watch --watch.onStart "echo start" --watch.onBundleStart "echo bundleStart" --watch.onBundleEnd "echo bundleEnd" --watch.onEnd "echo onEnd" --watch.onError "echo onError"',
```
Comment (lukastaegert) [inline test/cli/samples/watch/watch-event-hooks/_config.js:6]: `-cw` includes `--watch` 😉

File: test/cli/samples/watch/watch-event-hooks/_config.js (hunk 1 above)
Comment (lukastaegert) [reply test/cli/samples/watch/watch-event-hooks/_config.js:15]: Another idea: Split the output by line, filtering empty ones and mapping the remaining ones to their first word. Then you would just need to compare

```
start
bundles
bundleStart
created
...
```

File: test/cli/samples/watch/watch-event-hooks/_config.js (hunk 2 above)
Comment (Harris-Miller) [reply test/cli/samples/watch/watch-event-hooks/_config.js:6]: lol oops.

I could use some guidance on how to do this test exactly. The `test/cli/samples/` suite of test all seem to do tests on either console output or bundle file output. the one I threw up there was just a POC to test all of the watch event hooks at once by just looking for the echos being written

I can get that test working or I can take another direction. Would you like me to break up each hook into its own test? should I test that `--silent` doesn't print to stderr?

Comment (lukastaegert) [reply test/cli/samples/watch/watch-event-hooks/_config.js:6]: No, I think the test is fine in principle, and testing all hooks together seems adequate to me considering me that there is not too much logic involved. You should just make sure to make it a little more robust, e.g. see my other suggestion of preprocessing the output by splitting it into lines, doing some mapping and filtering and possibly doing an `deepStrictEqual` on the result. Looks to me like "echo" should also work on Windows, which is definitely a bonus.

**Comment by Harris-Miller** [comment]
I have 2 passing tests (had to test `onError` separately). They both take a very long time though
```
✔ watch-event-hooks: event hook shell commands write to stderr (40014ms)
```
I added in this that I found in a similar test file, though it doesn't help:
```javascript
	abortOnStderr(data) {
		if (data.includes('waiting for changes')) {
			return true;
		}
	}
```
What am I doing wrong?

**Comment by lukastaegert** [comment]
Looks like a process is not terminating correctly and needs to be force-quit by the test framework after the 40s timeout

**Comment by Harris-Miller** [comment]
I figured out the missing piece for the testing not exiting on `abortOnStderr`. All good now. I think this completes the requirements. Let me know if you need anything else

**Review by lukastaegert — APPROVED** [review]
Looks great! One last thing I wondered about is error handling, but maybe in watch mode to just print errors but not abort the build should be fine.

**Pull request merged**